### PR TITLE
Lightning: support charset conversion from GBK to UTF-8 for the CSV file

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Part of https://github.com/pingcap/br/issues/1389.

Support charset conversion from GBK to UTF-8 for the CSV file.

### What is changed and how it works?

Introduce `CharsetConvertor` in the CSV file parser.

### Check List 

Tests 

- [x] Unit test
- [x] Integration test

Side effects

- [x] Performance regression: consumes more Memory and CPU.

### Release note 

```release-note
Lightning: support charset conversion from GBK to UTF-8 for the CSV file.
```

### CODE REVIEW

**Comment by kennytm** [comment]
For future reference (no necessary action needed for this PR): GB18030 and GBK are not self-synchronizing, and all non-Unicode MBCS are not designed to be self-synchronizing.

When the `separator`, `delimiter` or `terminator` is a single character `a-z` `A-Z` ``@[\]^_`{|}~`` (U+0040–007E), the CSV parser can cut a character into half and produce an invalid field. Additionally, for GB18030 extends the set with `0-9` (U+0030–0039). 

For example, suppose `separator` is `|` (7C) and the input is `東西` (96 **7C** CE F7) it will be split into "\x96", "\xCE\xF7".

There are also non-trivial cases like `；` (A3 BB) can be found in the middle of `停火` (CD **A3 BB** F0).

We have similar issues for Big5 and Shift-JIS (U+0040–007F). 
EUC-KR seems fine for the ASCII separators but non-trivial cases can still happen (`，` (A3 AC) is inside `간Б` (B0 **A3 AC** A2)).

**Review by Little-Wallace — COMMENTED** [review]
Almost LGTM. But I am not familiar with the encoding format.

**Comment by glorv** [comment]
I'm not sure if we place the charset decode before the csv decoder instead of after it, then are we be able to avoid the problems in @kennytm 's comment? 🤔

BTW, since the charset is for the whole file not only for strings, I perfer to first do the convert to UTF8, and then parse the CSV.

**Comment by JmPotato** [comment]
> I'm not sure if we place the charset decode before the csv decoder instead of after it, then are we be able to avoid the problems in @kennytm 's comment? 🤔
> 
> BTW, since the charset is for the whole file not only for strings, I perfer to first do the convert to UTF8, and then parse the CSV.

Actually, the first PR tries to https://github.com/pingcap/tidb/pull/27172 do the conversion before parsing, however, we find that the correspondence of file position under two different encodings is difficult to maintain, i.e, we can not seek the origin GBK file with a UTF-8 byte index, so the chunk splitting could not work anymore. Converting the whole file before parsing is also not reasonable because the input file may be very large. After all of these problems, we decide to use this PR's way to do the conversion for the CSV file.

File: br/pkg/lightning/mydump/charset_convertor.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		return BINARY, dataInvalidCharReplace, errors.Errorf("found unsupported data-character-set: %s", cfg.DataCharacterSet)
+	}
+}
+
+var utf8RuneErrorBytes = []byte(string(utf8.RuneError))
+
+// Decode does the charset conversion work from sourceCharacterSet to utf8mb4.
+// It will return a byte slice as the conversion result whose length may be less or greater
+// than the original byte slice `src`.
+func (cc *CharsetConvertor) Decode(src []byte) ([]byte, error) {
```
Comment (kennytm) [inline br/pkg/lightning/mydump/charset_convertor.go:103]: use the `(*encoding.Decoder).String()` method for this.

File: br/pkg/lightning/mydump/charset_convertor.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		return false
+	}
+	return true
+}
+
+// GBKMax is the maximum number of bytes of a GBK encoded character.
+const GBKMax = 2
+
+// Encode will encode the data from utf8mb4 to sourceCharacterSet.
+func (cc *CharsetConvertor) Encode(src []byte) ([]byte, error) {
```
Comment (kennytm) [inline br/pkg/lightning/mydump/charset_convertor.go:136]: ditto, `(*encoding.Encoder).String()`.

**Review by glorv — APPROVED** [review]

**Pull request merged**